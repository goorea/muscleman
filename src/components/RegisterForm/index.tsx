import dayjs from 'dayjs';
import React, { useState } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { UseFormGetValues } from 'react-hook-form';
import DatePicker from 'react-native-date-picker';

import KeyboardAvoidingScrollView from '@src/components/KeyboardAvoidingScrollView';
import Text from '@src/components/Text';
import { useTheme } from '@src/contexts/ThemeProvider';
import { Gender } from '@src/types/graphql';

import useRenders from './hooks/useRenders';
import useRules from './hooks/useRules';
import {
  Container,
  Submit,
  Title,
  GenderContainer,
  GenderSelector,
  GenderSelectors,
} from './styled';

export type RegisterFormElement = {
  getValues: UseFormGetValues<RegisterUserFormInput>;
};

type P = {
  onSubmit: (input: RegisterUserFormInput) => void;
  errorMessages: string[];
};

export type RegisterUserFormInput = {
  name: string;
  nickname: string;
  tel: string;
  birth: string;
  gender: Gender;
};

const RegisterForm: React.ForwardRefRenderFunction<RegisterFormElement, P> = (
  { onSubmit, errorMessages },
  ref,
) => {
  const { colors } = useTheme();
  const {
    control,
    handleSubmit,
    setValue,
    watch,
    getValues,
    formState: { isSubmitting },
  } = useForm<RegisterUserFormInput>({
    defaultValues: { gender: Gender.Male },
  });
  const [visibleBirthPicker, setVisibleBirthPicker] = useState<boolean>(false);
  const { nameRules, nicknameRules, telRules, birthRules } = useRules();
  const { nameRender, nicknameRender, telRender, birthRender } = useRenders(
    setVisibleBirthPicker,
  );
  const onBirthPick = (date: Date) => {
    setVisibleBirthPicker(false);
    setValue('birth', dayjs(date).format('YYYY-MM-DD'));
  };
  const onBirthCancel = () => setVisibleBirthPicker(false);
  const setMale = () => setValue('gender', Gender.Male);
  const setFemale = () => setValue('gender', Gender.Female);

  React.useImperativeHandle(ref, () => ({
    getValues,
  }));

  return (
    <>
      <KeyboardAvoidingScrollView
        floatChildren={
          <Submit
            onPress={handleSubmit(onSubmit)}
            title="확인"
            disabled={isSubmitting}
          />
        }>
        <Container>
          <Title size={20} weight="bold">
            정보를 입력해주세요
          </Title>

          <Controller
            render={nameRender}
            name="name"
            control={control}
            rules={nameRules}
          />
          <Controller
            render={nicknameRender}
            name="nickname"
            control={control}
            rules={nicknameRules}
          />
          <Controller
            render={telRender}
            name="tel"
            control={control}
            rules={telRules}
          />
          <Controller
            render={birthRender}
            name="birth"
            control={control}
            rules={birthRules}
          />
          <GenderContainer>
            <Text color="grey3">성별</Text>

            <GenderSelectors>
              <GenderSelector
                title="남자"
                color={watch('gender') === Gender.Male ? 'primary' : 'disabled'}
                titleColor={watch('gender') === Gender.Male ? 'white' : 'black'}
                weight="normal"
                onPress={setMale}
              />
              <GenderSelector
                title="여자"
                color={
                  watch('gender') === Gender.Female ? 'primary' : 'disabled'
                }
                titleColor={
                  watch('gender') === Gender.Female ? 'white' : 'black'
                }
                weight="normal"
                onPress={setFemale}
              />
            </GenderSelectors>
          </GenderContainer>

          {errorMessages.map(error => (
            <Text size={12} color="error">
              {error}
            </Text>
          ))}
        </Container>
      </KeyboardAvoidingScrollView>

      <DatePicker
        modal
        mode="date"
        locale="ko"
        textColor={colors.foreground}
        open={visibleBirthPicker}
        date={dayjs(watch('birth')).toDate()}
        title="생년월일"
        confirmText="확인"
        cancelText="취소"
        onConfirm={onBirthPick}
        onCancel={onBirthCancel}
      />
    </>
  );
};

export default React.forwardRef(RegisterForm);
