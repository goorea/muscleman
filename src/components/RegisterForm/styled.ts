import styled from 'styled-components/native';
import Button from '@src/components/Button';
import Text from '@src/components/Text';

export const Container = styled.View`
  padding: 20px;
`;

export const Title = styled(Text)`
  margin-bottom: 20px;
`;

export const Submit = styled(Button)`
  margin-top: auto;
`;

export const GenderContainer = styled.View`
  margin-top: 20px;
`;

export const GenderSelectors = styled.View`
  flex-direction: row;
  margin-top: 10px;
`;

export const GenderSelector = styled(Button)`
  flex: 1;
  padding: 20px 0;
`;
