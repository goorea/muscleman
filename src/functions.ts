import StackFlashMessage, { Options } from 'react-native-stack-flash-message';

import { WeightSet } from '@src/types';
import { Set, TrainingType } from '@src/types/graphql';

export const flash = (options: Options) => StackFlashMessage.show(options);

export const getTrainingTypeForKorean = (type: TrainingType): string => {
  switch (type) {
    case TrainingType.Lower:
      return '하체';
    case TrainingType.Chest:
      return '가슴';
    case TrainingType.Back:
      return '등';
    case TrainingType.Shoulder:
      return '어깨';
    case TrainingType.Arm:
      return '팔';
    case TrainingType.Abdominal:
      return '복근';
    case TrainingType.Cardiovascular:
      return '유산소';
    default:
      return '기타';
  }
};

export const isWeightSets = (sets: Set[]): sets is WeightSet[] =>
  sets.every(
    set =>
      (set as WeightSet).weight !== undefined &&
      (set as WeightSet).count !== undefined,
  );
