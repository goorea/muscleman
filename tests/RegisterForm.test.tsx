import { act, fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import RegisterForm from '@src/components/RegisterForm';
import { wrapper } from '@tests/functions';

describe('RegisterForm 컴포넌트', () => {
  const onSubmit = jest.fn();
  const rendered = () =>
    render(<RegisterForm onSubmit={onSubmit} errorMessages={[]} />, {
      wrapper,
    });

  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = rendered();

    expect(toJSON()).toMatchSnapshot();
  });

  it('이름, 닉네임, 전화번호, 생일 입력 필드와 확인 버튼이 보인다', () => {
    const { queryByTestId, queryByText } = rendered();

    expect(queryByTestId('nameField')).not.toBeNull();
    expect(queryByTestId('nicknameField')).not.toBeNull();
    // expect(queryByTestId('telField')).not.toBeNull();
    // expect(queryByTestId('birthField')).not.toBeNull();
    expect(queryByText('확인')).not.toBeNull();
  });

  it('이름을 입력하지 않거나 2글자보다 적거나 8글자보다 많으면 에러메세지가 보인다', async () => {
    const { getByTestId, getByText, queryByText } = rendered();
    const nameField = getByTestId('nameField');

    expect(queryByText('이름을 입력해주세요')).toBeNull();
    expect(queryByText('2글자보다 적습니다')).toBeNull();
    expect(queryByText('8글자보다 많습니다')).toBeNull();

    await act(async () => fireEvent.press(getByText('확인')));

    expect(queryByText('이름을 입력해주세요')).not.toBeNull();

    await act(async () => fireEvent.changeText(nameField, '1'));

    expect(queryByText('2글자보다 적습니다')).not.toBeNull();

    await act(async () => fireEvent.changeText(nameField, '1'.repeat(9)));

    expect(queryByText('8글자보다 많습니다')).not.toBeNull();
  });

  it('닉네임을 입력하지 않거나 2글자보다 적거나 8글자보다 많으면 에러메세지가 보인다', async () => {
    const { getByTestId, getByText, queryByText } = rendered();
    const nicknameField = getByTestId('nicknameField');

    expect(queryByText('닉네임을 입력해주세요')).toBeNull();
    expect(queryByText('2글자보다 적습니다')).toBeNull();
    expect(queryByText('8글자보다 많습니다')).toBeNull();

    await act(async () => fireEvent.press(getByText('확인')));

    expect(queryByText('닉네임을 입력해주세요')).not.toBeNull();

    await act(async () => fireEvent.changeText(nicknameField, '1'));

    expect(queryByText('2글자보다 적습니다')).not.toBeNull();

    await act(async () => fireEvent.changeText(nicknameField, '1'.repeat(9)));

    expect(queryByText('8글자보다 많습니다')).not.toBeNull();
  });

  // it('휴대폰 번호를 입력하지 않거나 생일을 입력하지 않으면 에러 메세지가 보인다', async () => {
  //   const { getByText, queryByText } = rendered();
  //
  //   expect(queryByText('휴대폰 번호를 입력해주세요')).toBeNull();
  //   expect(queryByText('생년월일을 선택해주세요')).toBeNull();
  //
  //   await act(async () => fireEvent.press(getByText('확인')));
  //
  //   expect(queryByText('휴대폰 번호를 입력해주세요')).not.toBeNull();
  //   expect(queryByText('생년월일을 선택해주세요')).not.toBeNull();
  // });

  it('올바른 값을 입력한 뒤에 확인 버튼을 누르면 onSubmit 함수가 실행된다', async () => {
    const { getByTestId, getByText } = rendered();

    await act(async () => {
      fireEvent.changeText(getByTestId('nameField'), 'John');
      fireEvent.changeText(getByTestId('nicknameField'), 'John');
      // fireEvent.changeText(getByTestId('telField'), '01012341234');
      // fireEvent.changeText(getByTestId('birthField'), '1970-01-01');
    });

    await act(async () => fireEvent.press(getByText('확인')));

    expect(onSubmit).toHaveBeenCalled();
  });
});
