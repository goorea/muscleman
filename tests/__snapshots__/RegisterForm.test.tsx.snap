// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm 컴포넌트 렌더링이 올바르게 된다 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <Text
          italic={false}
          size={20}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "SpoqaHanSansNeo-Bold",
                "fontSize": 20,
                "fontStyle": "normal",
              },
              Object {
                "marginBottom": 20,
              },
            ]
          }
          weight="bold"
        >
          정보를 입력해주세요
        </Text>
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "baseline",
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 4,
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              collapsable={false}
              color="foreground"
              nativeID="animatedComponent"
              onLayout={[Function]}
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "SpoqaHanSansNeo-Regular",
                  "fontSize": 16,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 16,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              이름
            </Text>
            <TextInput
              autoCompleteType="name"
              autoFocus={true}
              label="이름"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="next"
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "SpoqaHanSansNeo-Regular",
                    "fontSize": 16,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": "100%",
                  },
                ]
              }
              testID="nameField"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "baseline",
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 4,
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              collapsable={false}
              color="foreground"
              nativeID="animatedComponent"
              onLayout={[Function]}
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "SpoqaHanSansNeo-Regular",
                  "fontSize": 16,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 16,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              닉네임
            </Text>
            <TextInput
              autoCompleteType="username"
              label="닉네임"
              name="nickname"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="next"
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "SpoqaHanSansNeo-Regular",
                    "fontSize": 16,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": "100%",
                  },
                ]
              }
              testID="nicknameField"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#2687ED",
        "justifyContent": "center",
        "marginTop": "auto",
        "opacity": 1,
        "paddingBottom": 14,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 14,
      }
    }
  >
    <Text
      color="white"
      italic={false}
      size={16}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontFamily": "SpoqaHanSansNeo-Bold",
            "fontSize": 16,
            "fontStyle": "normal",
          },
        ]
      }
      weight="bold"
    >
      확인
    </Text>
  </View>
</View>
`;
